### PR TITLE
Issue 16736 - Retrieving cUrl time values is quite cumbersome

### PR DESCRIPTION
**Pull request by RazvanN7**
I added a function which is a wrapper over the C api function curl_easy_getinfo in order to ease the fetching of various timings. I only tested for http that the c function is called correctly (i.e. after the function is called the result is CURLE_OK) since I do not know how to test if the timings are correct.

### CODE REVIEW

File: std/net/curl.d [hunk 1]
```diff
@@ -2076,6 +2076,7 @@ private mixin template Protocol()
         http.onReceive = (ubyte[] data) { return data.length; };
         http.setAuthentication("user", "pass");
         http.perform();
+
```
Comment (JackStouffer) [inline std/net/curl.d]: remove

Comment (RazvanN7) [inline std/net/curl.d]: Are you talking about the whole unittest? This wasn't my addition

**Comment by RazvanN7** [comment]
@JackStouffer is this ok?

File: std/net/curl.d (hunk 1 above)
Comment (JackStouffer) [inline std/net/curl.d]: the extra line

File: std/net/curl.d [hunk 2]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
         p.curl.set(CurlOption.useragent, userAgent);
     }
 
+    /**
+     * Get various timings defined in etc.c.curl.CurlInfo.
```
Comment (andralex) [inline std/net/curl.d]: Everything that's not English prose should be formatted specially, in this case as code - backticks for single words, `$(D ...)` for multiple words.

File: std/net/curl.d [hunk 3]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
...
+     * The value is usable only if the return value is equal to CurlError.ok.
```
Comment (andralex) [inline std/net/curl.d]: format

File: std/net/curl.d [hunk 4]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
...
+     *
+     * Params:
+     *      timing = one of the timings defined in etc.c.curl.CurlInfo.
```
Comment (andralex) [inline std/net/curl.d]: format (only "etc.c.curl.CurlInfo", "timing" will be formatted automatically)

File: std/net/curl.d [hunk 5]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
...
+     *               The values are: namelookup_time, connect_time, pretransfer_time,
```
Comment (andralex) [inline std/net/curl.d]: format

File: std/net/curl.d [hunk 6]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
...
+     *               starttransfer_time, redirect_time, appconnect_time, total_time.
```
Comment (andralex) [inline std/net/curl.d]: format

File: std/net/curl.d [hunk 7]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
...
+     *      val    = the actual value of the inquired timing
+     *
+     * Returns:
+     *      The return code of the operation. The value stored in val
```
Comment (andralex) [inline std/net/curl.d]: format

File: std/net/curl.d [hunk 8]
```diff
@@ -2836,6 +2836,40 @@ struct HTTP
...
+     *      should be used only if the return value is CurlError.ok.
```
Comment (andralex) [inline std/net/curl.d]: format

File: std/net/curl.d [hunk 9]
```diff
@@ -3480,6 +3533,16 @@ struct FTP
...
     }
+
+    /**
+       Get the various timings like name lookup time, total time, connect time etc.
+       The timed category is passed through the timing parameter while the timing
+       value is stored at val. The value is usable only if res is equal to CurlError.ok.
+    */
+    CurlCode getTiming(CurlInfo timing, ref double val)
+    {
+        return p.curl.getTiming(timing, val);
```
Comment (andralex) [inline std/net/curl.d]: not covered, wonder why because I see calls in the unittest above

Comment (andralex) [inline std/net/curl.d]: ah, only the other overload is used - please unittest this as well

**Comment by andralex** [comment]
@JackStouffer in order to improve flow please follow through after you request changes, or submit them as comments. I'll dismiss your review for now. Thx!

File: std/net/curl.d (hunk 2 above)
Comment (RazvanN7) [inline std/net/curl.d]: Multiple words with linked_by_underscore are formatted with backticks or $(D ...)?

File: std/net/curl.d (hunk 9 above)
Comment (RazvanN7) [inline std/net/curl.d]: I am certainly missing something, but I can't seem to find the FTP unittests. All unittests only test for the HTTP methods and classes - I see that there is a HTTP server implemented in a unittest but I do not see any ftp implemented server. On the other hand the FTP code seem to be covered, except for my method. I have no idea how that code is covered...

**Comment by JackStouffer** [comment]
In most of the formatting cases that Andrei pointed out, you should actually be using `$(LREF ...)` or  `$(REF ...)`. These create links in the docs, which you should use whenever you're referencing other functions or values in the standard lib.

**Comment by RazvanN7** [comment]
@andralex @JackStouffer is this ok?

File: std/net/curl.d [hunk 10]
```diff
@@ -3480,6 +3534,66 @@ struct FTP
     {
         p.curl.set(CurlOption.infilesize_large, to!curl_off_t(len));
     }
+
+    /**
+     * Get various timings defined in $(LREF etc.c.curl.CurlInfo).
```
Comment (JackStouffer) [inline std/net/curl.d]: should be `$(REF CurlInfo, etc, c, curl)`

File: std/net/curl.d [hunk 11]
```diff
@@ -3480,6 +3534,66 @@ struct FTP
...
+     * The value is usable only if the return value is equal to $(LREF CurlError.ok).
```
Comment (JackStouffer) [inline std/net/curl.d]: Broken link, as `CurlError` is not defined in the same file. Should be `$(REF ok, etc, c, curl, CurlError)`

**Comment by JackStouffer** [comment]
Sorry, I should have been more clear. `LREF` is a shortcut for referencing functions in the same file, `REF` is for everything else.

**Comment by RazvanN7** [comment]
@JackStouffer Thanks for the clarifications. It should be ok now

**Comment by JackStouffer** [comment]
Ok, now I'm confused because that doesn't work either. It looks like `REF` doesn't work for enum members.

For now, switch all `REF`s that point to `enum` members to just be wrapped `$(D )`.

Sorry to have you run around with this! Just ping me again when you're done and I'll pull.

**Comment by JackStouffer** [comment]
BTW you can look at the generated docs for your PR by clicking on the "Details" link next to "CyberShadow/DAutoTest" in the checks section.

**Comment by RazvanN7** [comment]
@JackStouffer No probs. I nested all the enum members with ($D ...). Thanks for reviewing!

**Comment by JackStouffer** [comment]
Auto-merge toggled on

**Pull request merged**